syntax = "proto3";

enum Side {
  LEFT = 0;
  RIGHT = 1;
  SIDE_COUNT = 2;
}

message EngCtl {
  Side side = 1;
  oneof method {
    float pwm = 2;
    float speed = 3;
  }
}

message PIDConfig {
  float p = 1;
  float i = 2;
  float d = 3;
}

message PIDReset {}

message WheelState {
  uint64 timestamp = 1;
  Side side = 2;
  float pwm = 3;
  float speed = 4;
  float mileage = 5;
  float targetSpeed = 6;
}

message PIDState {
  uint64 timestamp = 1;
  float error = 2;
  float integral = 3;
  float pOut = 4;
  float iOut = 5;
  float dOut = 6; 
}

message KitekMsg {
  oneof msg {
    EngCtl engCtl = 1;
    PIDReset resetPID = 2;
    PIDConfig cfgPID = 3;
    WheelState wheelState = 4;
    PIDState pidState = 5;
  }
}

// EmitUDP.cpp
message Point2D {
  float x = 1;
  float y = 2;
}